Test failure (27138) TestGossipFirstRange

The test failed due to a timeout: https://teamcity.cockroachdb.com/viewQueued.html?itemId=27138

```
[...]
W161005 11:10:37.335823 20922 ts/db.go:113 [n3,ts-poll] node unavailable; try another peer W161005 11:10:42.247232 20727 gossip/infostore.go:295 [n1] node unavailable; try another peer I161005 11:10:42.250704 20674 kv/txn_coord_sender.go:251 [n1] txn coordinator: 0.01 txn/sec, 100.00/0.00/0.00/0.00 %cmmt/cmmt1pc/abrt/abnd, 0s/0s/0s avg/σ/max duration, 0.0/0.0/0 avg/σ/max restarts (0 samples) 
I161005 11:10:42.265809 20757 kv/txn_coord_sender.go:251 [n2] txn coordinator: 0.00 txn/sec, 100.00/0.00/0.00/0.00 %cmmt/cmmt1pc/abrt/abnd, 0s/0s/0s avg/σ/max duration, 0.0/0.0/0 avg/σ/max restarts (0 samples) 
W161005 11:10:42.297219 20747 gossip/infostore.go:295 [n2] node unavailable; try another peer 
I161005 11:10:42.324938 20823 kv/txn_coord_sender.go:251 [n3] txn coordinator: 0.00 txn/sec, 100.00/0.00/0.00/0.00 %cmmt/cmmt1pc/abrt/abnd, 0s/0s/0s avg/σ/max duration, 0.0/0.0/0 avg/σ/max restarts (0 samples) 
W161005 11:10:42.329529 20919 gossip/infostore.go:295 [n3] node unavailable; try another peer 

panic: test timed out after 2m0s 
goroutine 21110  [running]: 
panic(0x1619600, 0xc420ec80a0) 
/usr/local/go/src/runtime/panic.go:500 +0x1a1 testing.startAlarm.func1() 
/usr/local/go/src/testing/testing.go:918 +0x10b created by time.goFunc 
/usr/local/go/src/time/sleep.go:154 +0x44
[...]
```

--- a/storage/gossip_test.go
+++ b/storage/gossip_test.go
@@ -42,7 +42,7 @@ func TestGossipFirstRange(t *testing.T) {
 
 	errors := make(chan error)
 	descs := make(chan *roachpb.RangeDescriptor)
-	tc.Servers[0].Gossip().RegisterCallback(gossip.KeyFirstRangeDescriptor,
+	unregister := tc.Servers[0].Gossip().RegisterCallback(gossip.KeyFirstRangeDescriptor,
 		func(_ string, content roachpb.Value) {
 			var desc roachpb.RangeDescriptor
 			if err := content.GetProto(&desc); err != nil {
@@ -50,7 +50,13 @@ func TestGossipFirstRange(t *testing.T) {
 			} else {
 				descs <- &desc
 			}
-		})
+		},
+	)
+	// Unregister the callback before attempting to stop the stopper to prevent
+	// deadlock. This is still flaky in theory since a callback can fire between
+	// the last read from the channels and this unregister, but testing has
+	// shown this solution to be sufficiently robust for now.
+	defer unregister()
 
 	// Wait for the specified descriptor to be gossiped for the first range. We
 	// loop because the timing of replica addition and lease transfer can cause

